pkg/scanners/scannerv4/nodescan_convert.go

```diff
@@ -1,76 +1,93 @@
 	"github.com/stackrox/rox/generated/storage"
-	"github.com/stackrox/rox/pkg/clair"
-	"github.com/stackrox/rox/pkg/cvss/cvssv3"
 	"github.com/stackrox/rox/pkg/protocompat"
+	"github.com/stackrox/rox/pkg/set"
+	"github.com/stackrox/rox/pkg/utils"
+)
+
+var (
+	rhcosOSImageRegexp = regexp.MustCompile(`(Red Hat Enterprise Linux) (CoreOS) ([\d])([\d]+)`)
 )
 
-func ToNodeScan(r *v4.VulnerabilityReport) *storage.NodeScan {
+func toNodeScan(r *v4.VulnerabilityReport, osImageRef string) *storage.NodeScan {
+	// TODO(ROX-26593): Instead of fixing labels here, add RHCOS DistributionScanner to ClairCore
+	fixedNotes := fixNotes(toStorageNotes(r.Notes), osImageRef)
+
 	return &storage.NodeScan{
-		ScanTime:       protocompat.TimestampNow(),
-		Components:     toStorageComponents(r),
-		Notes:          toStorageNotes(r.Notes),
-		ScannerVersion: storage.NodeScan_SCANNER_V4,
+		ScanTime:        protocompat.TimestampNow(),
+		Components:      toStorageComponents(r),
+		Notes:           fixedNotes,
+		ScannerVersion:  storage.NodeScan_SCANNER_V4,
+		OperatingSystem: toOperatingSystem(osImageRef),
 	}
 }
 
+func toOperatingSystem(ref string) string {
+	r := rhcosOSImageRegexp.FindStringSubmatch(ref)
+	if len(r) != 5 {
+		return ""
+	}
+	return fmt.Sprintf("rhcos:%s.%s", r[3], r[4])
+}
+
 func toStorageComponents(r *v4.VulnerabilityReport) []*storage.EmbeddedNodeScanComponent {
-	result := make([]*storage.EmbeddedNodeScanComponent, 0)
-	vulnerabilities := r.GetVulnerabilities()
 	packages := r.GetContents().GetPackages()
+	result := make([]*storage.EmbeddedNodeScanComponent, 0, len(packages))
+
+	for _, pkg := range packages {
+		vulns := getPackageVulns(pkg.GetId(), r)
+		result = append(result, createEmbeddedComponent(pkg, vulns))
+	}
+	return result
+}
 
-	for pkgID, pkgvuln := range r.GetPackageVulnerabilities() {
-		pkg := findPackage(packages, pkgID)
-		if pkg == nil {
-			log.Warnf("Unable to find package for id %s in %d packages. Skipping", pkgID, len(packages))
+func getPackageVulns(packageID string, r *v4.VulnerabilityReport) []*storage.EmbeddedVulnerability {
+	vulns := make([]*storage.EmbeddedVulnerability, 0)
+	mapping, ok := r.GetPackageVulnerabilities()[packageID]
+	if !ok {
+		// No vulnerabilities for this package, skip
+		return vulns
+	}
+	processedVulns := set.NewStringSet()
+	for _, vulnID := range mapping.GetValues() {
+		if !processedVulns.Add(vulnID) {
+			// Already processed this vulnerability, skip it
 			continue
 		}
-		vulns := make([]*storage.EmbeddedVulnerability, 0)
-		for _, vulnID := range pkgvuln.GetValues() {
-			vuln, ok := vulnerabilities[vulnID]
-			if !ok {
-				log.Warnf("Unable to find vulnerability for id %s in %d vulnerabilites. Skipping", vulnID, len(vulnerabilities))
-				continue
-			}
-			vulns = append(vulns, convertVulnerability(vuln))
+		vulnerability, ok := r.Vulnerabilities[vulnID]
+		if !ok {
+			log.Warnf("Unable to find vulnerability %s in report - skipping this vulnerability", vulnID)
+			continue
```

Comment:
Two thoughts:
1. The mapping comes from the report, so the wording that this is unable to find the vuln in the report may be not fully accurate.
2. This warning will be displayed to the user - are we sure that the user would know what does it mean to skip?

So maybe let's rephrase this a bit - you could propose something or I could help, but for the latter I would need to understand what is the difference between `r.GetPackageVulnerabilities()[packageID]` and `r.Vulnerabilities`

Additionally, if this warning will be shown, then it may flood the logs, as I expect this will happen for all the rescans for the nodes that are roughly on the same version. Maybe we should reduce the chattiness?
An alternative would be to return a second slice here with something like `skippedVulns` and then log it in the caller function in one line.

Reply:
1. That is correct. Technically, the mapping is inaccurate in this case.
2. I don't think this is actionable for the user. If the mapping is faulty, there is nothing a user can do about it. In fact, this is more of a safety check to ensure we don't try to access out of bounds of the map. 

Especially as it's not actionable for a user *and* would likely lead to a lot of bug reports when read by users, I've bumped it down to debug.

> but for the latter I would need to understand what is the difference between r.GetPackageVulnerabilities()[packageID] and r.Vulnerabilities

Let me explain that with the example [from the tests here](https://github.com/stackrox/stackrox/blob/b99d5cb46cd54b1d1257178d812824492061abc8/pkg/scanners/scannerv4/nodescan_convert_test.go#L197-L218):
The report contains 
* a list of `Vulnerabilities`, which represent all the CVE data of a given vuln
* `Contents.Packages`, which are *all installed packages* we discovered on the node
* `PackageVulnerabilities`, which connects the two data structures above

The `PackageVulnerabilities` *Key* is the ID of a `Contents.Package`. The slice of strings of values are the IDs of all `Vulnerabilities` that this given package has.
If the mapping were to contain `"0": {Values: []string{"7401229"}},`, that would mean that package with ID 0 has 1 vulnerability with ID 7401229. 
This safeguard we are talking about ensures that the vulnerability 7401229 really exists in the `Vulnerabilities` map. If it doesn't, that means that either part of the report was corrupted (bad), or that the Scanner  v4 Matcher generated a broken report (worse). I deem both scenarios rather unlikely.

Reply:
Okay, I think I understand now. Then, the message should say something like:

```go
log.Warnf("Package %s contains a vulnerability with unknown ID %s. This vulnerability won't be stored, packageID, vulnID)
```
Does that make sense?

Reply:
I like it! I've rephrased it slightly to 
`Mapping for package %s contains a vulnerability with unknown ID %s. This vulnerability won't be stored`

Reply:
fantastic!